### ISSUE TITLE
Addon Manager git path issues on macOS

### ISSUE THREAD

**Issue by JohnCulver** [LuaLS/vscode-lua#81]
### Which OS are you using?

MacOS

### Expected Behaviour

The plugin should function after install

### Actual Behaviour

...

I also see the following error:

<details><summary>Retrieved Log</summary>

```
#### STARTUP ####
[2023-03-02 11:20:40] |   DEBUG   |     Filesystem     | Created directory at "/Users/me/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager"
[2023-03-02 11:20:40] |   ERROR   |         Git        | fatal: not a git repository (or any of the parent directories): .git


```

</details>

### Reproduction steps

Install the plugin

### Log

```
#### STARTUP ####
[2023-03-02 11:20:40] |   DEBUG   |     Filesystem     | Created directory at "/Users/me/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager"
[2023-03-02 11:20:40] |   ERROR   |         Git        | fatal: not a git repository (or any of the parent directories): .git


```

**Comment by flavius-popan** [comment]
This is super bizarre, within the addonManager directory is a recursive copy of all folders leading up to itself before it finally lists files I'd expect to see in the directory

![image](https://user-images.githubusercontent.com/832651/222548495-001beb0f-0f91-4031-9f5a-1fa1e8c9f550.png)

**Comment by flavius-popan** [comment]
I copied the files inside the second level up to the first level, deleted the Users dir and all files from addonManager, and everything started up just fine with no errors. Seems like a recent code change recursively copied it's abs path within the plugin.

**Comment by carsakiller** [comment]
Extremely helpful, thank you! I believe it is, like you say, a pathing issue. It is necessary to chop off the leading slash on Windows, but it is causing this colossal mess on UNIX systems, it appears.

**Comment by flavius-popan** [comment]
Glad I could be of assistance, good luck!

**Comment by carsakiller** [comment]
I have pushed some commits that should resolve this issue. Thank you for the help!

If you have already run the addon manager, you will likely have to delete `/Users/me/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager`. Once the fix has been released, try the addon manager command again and it should be fixed 🙂

**Comment by carsakiller** [comment]
Looks like there are still issues on macOS even after the latest patch. I do not a mac to test on so it is extra challenging to try and resolve this issue. Is someone able to provide additional details on why the macOS path would differ from the Linux path when using [`Uri.fsPath`](https://code.visualstudio.com/api/references/vscode-api#Uri)?

**Comment by carsakiller** [comment]
I have added some additional logging. Could someone with a mac please install the vsix from the [latest action run](https://github.com/LuaLS/vscode-lua/actions/runs/4328437923) and let me know what path is logged to the log file?

![image](https://user-images.githubusercontent.com/61925890/222863405-fde0ede9-663b-43fa-b589-0ece4e4a0469.png)

**Comment by Rhys-T** [comment]
(Joining in from #88.)

Assuming I have the right log file, it looks almost the same as before. The only difference I see is that the `Failed to checkout repo!` I was getting on the middle line has become `Failed to pull LLS-Addons!`. 

```
#### STARTUP ####
[2023-03-03 20:02:54] |   DEBUG   |     Filesystem     | Created directory at "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager"
[2023-03-03 20:02:54] |   ERROR   |         Git        | Failed to pull LLS-Addons!
[2023-03-03 20:02:54] |   ERROR   |         Git        | fatal: not a git repository (or any of the parent directories): .git
```

Also, for some reason, the VSIX version of the extension seems to keep uninstalling itself. It keeps saying "Please reload Visual Studio Code to complete the uninstallation of this extension." even though I never told VSCode to remove it, and sure enough, when I let it reload the window, the Lua extension is gone entirely. Not sure what's going on there.

**Comment by carsakiller** [comment]
Try deleting `/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager`

**Comment by Rhys-T** [comment]
Trashed that folder, reloaded the window, and reinstalled from the VSIX file (downloaded from [build #262](https://github.com/LuaLS/vscode-lua/actions/runs/4328437923)). It seems to be working correctly this time, and cloned the repo into the right place. It also isn't uninstalling itself anymore.

**Comment by carsakiller** [comment]
Interesting, all I did was add more logging 😄. Maybe `3.6.15` broke that directory, and simply deleting it and letting the addon manager handle it fixed it.

Can another Mac user do the same? Delete that directory and install the vsix from the latest build-just to confirm.

**Comment by Rhys-T** [comment]
I don't currently have another macOS-level account set up (and running VSCode with a separate profile directory doesn't seem to give it a separate set of installed extensions), but I just did some more testing: deleted `addonManager`, launched VSCode (without any version of the Lua extension installed), and installed build \#262 from the .vsix… and it got an error again. But it looks a little different this time:

```
#### STARTUP ####
[2023-03-04 02:04:17] |   DEBUG   |     Filesystem     | Created directory at "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager"
[2023-03-04 02:04:17] |   ERROR   |         Git        | Failed to pull LLS-Addons!
[2023-03-04 02:04:17] |   ERROR   |         Git        | error: pathspec 'main' did not match any file(s) known to git
```

That's the first time I've seen the last line from Git be something other than `fatal: not a git repository (or any of the parent directories): .git`. I'm not entirely sure which directory it's trying to do a `git pull` in. I can't find anything called `LLS-Addons`, and the only `.git` directory I see anywhere inside `globalStorage/sumneko.lua` is directly inside `addonManager`.

**Comment by carsakiller** [comment]
That appears to be a different error, which is progress! I have hopefully just fixed that one as well but let's track that in another issue.

**Comment by carsakiller** [comment]
Is a Mac user able to confirm if this is working in `v3.6.17`?

**Comment by Rhys-T** [comment]
Just updated, and opened a Lua project. So far, it doesn't seem to be throwing those errors in the bottom-right corner anymore.

However, I opened the Lua Addon Manager, and got this alert (with the addon manager behind it):

> **Error**
> Failed to fetch addons! Please check your connection to GitHub.
> **[Okay]** [Open Log]

Looking at the log, this seems to be the main error message:
```
error: cannot lock ref 'refs/remotes/origin/main': is at d8e758ed0e94337a9b54295dadbf0077ed5e0001 but expected c5ba0df46a487ffee1bfd51dbe34ec1fca0a4554
```

I'm not sure whether:
- something is messed up in my existing `addonManager` directory as a result of the versions I was trying before, and it just needs to be reset again
- multiple windows were trying to do things with that directory at once
- something else entirely

If it helps, some of the results I'm finding for that error say that it can result from having two branches with the same name but different capitalization, or things like that. I'm on macOS, which defaults to making case-insensitive filenames, so if this is a macOS-specific problem, that may be why. On the other hand, it could be something completely unrelated.

I can remove that folder one more time and try again, but I wanted to wait in case you needed me to check anything else about it first.

<details>
<summary>Full log</summary>

```
#### STARTUP ####
[2023-03-09 10:20:17] |   DEBUG   |       WebVue       | Workspace Open: true
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Created directory at "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager"
[2023-03-09 10:20:17] |  VERBOSE  |       Command      | Executing "getAddons" ({"command":"getAddons","data":{"page":1}})
[2023-03-09 10:20:17] |   ERROR   |    Addon Manager   | Failed to fetch addons! Please check your connection to GitHub.
[2023-03-09 10:20:17] |   ERROR   |    Addon Manager   | error: cannot lock ref 'refs/remotes/origin/main': is at d8e758ed0e94337a9b54295dadbf0077ed5e0001 but expected c5ba0df46a487ffee1bfd51dbe34ec1fca0a4554
From https://github.com/LuaLS/LLS-Addons
 ! c5ba0df..d8e758e  main       -> origin/main  (unable to update local ref)

[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found busted
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found cc-tweaked
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found cocos4.0
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found love2d
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found lovr
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found luaecs
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found luafilesystem
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found luassert
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found luvit
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found moneymoney
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found openresty
[2023-03-09 10:20:17] |  VERBOSE  |    Addon Manager   | Found skynet
[2023-03-09 10:20:17] |    INFO   |        Addon       | Found update for "cc-tweaked"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/busted/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/cc-tweaked/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/love2d/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/cocos4.0/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/lovr/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/luassert/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/luaecs/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/luvit/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/moneymoney/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/luafilesystem/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/openresty/info.json"
[2023-03-09 10:20:17] |   DEBUG   |     Filesystem     | Read "/Users/██████/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager/addons/skynet/info.json"
[2023-03-09 10:21:52] |  VERBOSE  |       Command      | Executing "openLog" ({"command":"openLog"})
```

</details>

**Comment by carsakiller** [comment]
I can only guess what is causing this issue based off what I could find online. It seems to come down to what you have already mentioned. It should not be caused by similar branch names, there is only one branch, unless it is _somehow_ conflicting with a local branch?

I did not test how it behaves if there are multiple VS Code windows using the addon manager at once… that could be a problem I suppose. This should be reproducible on other OSes as well, though, so I can try this later.

Could you run `git branch -a` in `/Users/YOU/Library/Application Support/Code/User/globalStorage/sumneko.lua/addonManager` and share the results? You can maybe try [pruning origin](https://stackoverflow.com/a/63772837). Otherwise, try deleting again and let's hope it is just from the old versions.

**Comment by Rhys-T** [comment]
I only had the addon manager open in one window, and as far as I know, none of the other windows even had any Lua code in their folders. I just mentioned that because it said it was trying and failing to 'lock' something, and I did have other windows, which might conceivably have been trying to do something with that repo (I had just reloaded them, after updating the extension).

`git branch -a` returns:
```
* main
  remotes/origin/HEAD -> origin/main
  remotes/origin/main
```

I just closed and reopened the addon manager, without doing anything to the repo, and it didn't show that alert. I then reloaded the window again, opened the manager again, and still didn't get an alert. I also installed and then disabled another addon successfully. It may have just been a one-time glitch. I'll let you know if it happens again.
